language/vm/vm-runtime/src/libra_vm.rs

```diff
@@ -1,100 +1,550 @@
                 TransactionBlock::WriteSet(_) => unimplemented!(""),
             }
         }
+        report_block_count(count);
+        Ok(result)
+    }
+
+    fn execute_user_transactions(
+        &mut self,
+        txn_block: Vec<SignedTransaction>,
+        data_cache: &mut BlockDataCache<'_>,
+        state_view: &dyn StateView,
+    ) -> VMResult<Vec<TransactionOutput>> {
+        let signature_verified_block: Vec<Result<SignatureCheckedTransaction, VMStatus>> =
+            txn_block
+                .into_par_iter()
+                .map(|txn| {
+                    txn.check_signature()
+                        .map_err(|_| VMStatus::new(StatusCode::INVALID_SIGNATURE))
+                })
+                .collect();
+        let mut result = vec![];
+        for transaction in signature_verified_block {
+            record_stats! {time_hist | TXN_TOTAL_TIME_TAKEN | {
+                    let output = match transaction {
+                        Ok(txn) => self.execute_user_transaction(state_view, data_cache, &txn),
+                        Err(e) => discard_error_output(e),
+                    };
+                    report_execution_status(output.status());
 
+                    // `result` is initially empty, a single element is pushed per loop iteration and
+                    // the number of iterations is bound to the max size of `signature_verified_block`
+                    assume!(result.len() < usize::max_value());
+                    result.push(output);
+                }
+            }
+        }
         Ok(result)
     }
 }
 
+pub(crate) fn discard_error_output(err: VMStatus) -> TransactionOutput {
+    // Since this transaction will be discarded, no writeset will be included.
+    TransactionOutput::new(
+        WriteSet::default(),
+        vec![],
+        0,
+        TransactionStatus::Discard(err),
+    )
+}
+
+// Validators external API
+impl VMVerifier for LibraVM {
+    /// Determine if a transaction is valid. Will return `None` if the transaction is accepted,
+    /// `Some(Err)` if the VM rejects it, with `Err` as an error code. We verify the following
+    /// items:
+    /// 1. The signature on the `SignedTransaction` matches the public key included in the
+    ///    transaction
+    /// 2. The script to be executed is in the whitelist.
+    /// 3. Invokes `LibraAccount.prologue`, which checks properties such as the transaction has the
+    /// right sequence number and the sender has enough balance to pay for the gas.
+    /// 4. Transaction arguments matches the main function's type signature.
+    fn validate_transaction(
+        &self,
+        transaction: SignedTransaction,
+        state_view: &dyn StateView,
+    ) -> Option<VMStatus> {
+        let data_cache = BlockDataCache::new(state_view);
+        record_stats! {time_hist | TXN_VALIDATION_TIME_TAKEN | {
+                let gas_schedule = self.move_vm.load_gas_schedule(state_view, &data_cache);
```

Comment:
why not use the `CostTable` in `LibraVM`? I understand this was like this and something you told me about (and yes it's odd) but given we are loading the table in `LibraVM` why not use that? (I guess also the way we load it would have to be changed slightly).
Also do I understand it correctly we are loading the table for every single transaction we are validating?

Reply:
That is exactly the problem that I had. Notice the LibraVM::new doesn't take in a StateView. This means that the created LibraVM will have an empty gas schedule. Ideally there should be an `LibraVM::init(&mut self, state_view: &StateView)` to load the gas schedule from the AC side, but this is not how things are set up for now. If we just call `LibraVM::new` and then `LibraVM:: validate_transaction`, no gas schedule can be preserved (notice that `validate_transaction` takes in a `&self` rather than an `&mut self` which will prevent us from persisting the cost table)
> Also do I understand it correctly we are loading the table for every single transaction we are validating?

Yes, and this is what we used to do as well.

Reply:
yeah I agree with the `init` though it feels that we can change the API to take a `&mut self` or use some interior mutability?
In any case definitely something to fix in time